### ISSUE TITLE
Next Release?

### ISSUE THREAD

**Issue by esc** [graphite-project/graphite-web#677]
How does everyone feel about cutting a new release? I think it's high time, there are many new features and several bugs that have been fixed. The questions that nag me is how to release? I couldn't find any maintainers notes anywhere and there are many small details to be taken care of. Let me sketch some bullet-points from my previous experience releasing software, these can serve as a skeleton to be expanded.

- [x] mark any pull-requests / issues that should or need to be taken care of
- [x] take care of these pull-requests
- [x] update the release notes
- [x] update the changelog and commit to master
- [x] update the version number and and commit to ~~master~~ 0.9.x
- [x] tag the release (ideally using a signed Git tag)
- [x] push the master
- [x] push the release tag
- [x] bump version on master to next development version and push
- [x] upload to PyPI
- [x] generate documentation and upload (to where?)
- [x] new website (https://github.com/graphite-project/graphite-project.github.io/issues)
- [x] announce release (twitter?)

Some things to note:

* Only critical bug fixes and no new features will be merged to master as soon as we start cutting the release
* We can of course do the "release candidate" thing, where you cut two or three release candidates which you make available for testing for a week or so, that can help to stabilize the release.
* Do we want to release only graphite-web or also ceres/whisper/carbon?
* Lastly, the release procedure and all the things that were done will be written down and committed after the release to help with future releases.

**Comment by ojilles** [comment]
Just wanted to chime in and say thanks for all the hard work folks!

**Comment by jraby** [comment]
@obfuscurity, I plan on upgrading our clusters to 0.9.13 (+ our local patches) this week, I'll report back if I encounter any weirdness.

**Comment by obfuscurity** [comment]
Fixed documentation for pip installations in https://github.com/graphite-project/graphite-web/pull/1080.

**Comment by jraby** [comment]
For the record, we've been running 0.9.x + https://github.com/graphite-project/carbon/pull/351 + https://github.com/graphite-project/whisper/pull/105 without issue for a few days.

**Comment by timbunce** [comment]
Only [two issues](https://github.com/graphite-project/graphite-web/milestones/0.9.13) left open on the 0.9.13 milestone, and one of those is this one!
The other one, #1081, hasn't been updated in two months.
Is it really a blocker?
Is it the only one?

**Comment by obfuscurity** [comment]
Yes, there's not much left but those are clearly sticking points for a good release.

**Comment by deniszh** [comment]
I have an idea - as we have couple of issues with current 0.9.13-pre1 (like https://github.com/graphite-project/carbon/issues/383 or https://github.com/graphite-project/carbon/issues/355) when pypy decides that 0.9.13 is already released - maybe we need to rename current release to 0.9.14 then?

**Comment by deniszh** [comment]
Looks like https://github.com/graphite-project/graphite-web/pull/1188 is a good thing to have in upcoming release. Will try to port it.

**Comment by timbunce** [comment]
This case was opened on Apr 3, _2014_. I deeply appreciate the desire to create a good release but it seems that momentum keeps getting lost.

Please, just ship something! Then ship something a little better in a month or two. Then ship something a little better a month or two after that. Version numbers are not a finite resource and momentum breeds contributions.

Or perhaps it's time to admit defeat and, echoing @obfuscurity's [comment in June 2014](https://github.com/graphite-project/graphite-web/issues/677#issuecomment-46499955] projects like graphite-api and cyanite as the future of the Graphite community), tell people that "projects like graphite-api and cyanite are the future of the Graphite community" and that this project is near-dead.

**Comment by frejsoya** [comment]
Sorry for the 'me2' post, but just release something. Either there are  a lot out there just running master or all the new contributions are not being used.

**Comment by esc** [comment]
A 0.9.13-pre was released last december:

https://pypi.python.org/pypi/graphite-web/0.9.13

This issue should have been closed then, so closing now.

@timbunce and @frejsoya please try the latest available release from PyPi.

**Comment by nox** [comment]
That's a pre-release. Why can't we have a proper release?

**Comment by esc** [comment]
@nox you want me to reopen the issue?

**Comment by steve-dave** [comment]
We shouldn't have a "proper" release till we have a stable release.  We can't have a stable release until all the required fixes are developed, PRed, tested and merged.  Pre-release is the closest you'll get to a real release until there is confidence that the release will be solid.  I don't think you'll find anyone that doesn't want a release...  but it's harder to find people that have the time, willingness and ability to make it a reality.

If you're personally comfortable with the current state then just check out your choice branch and run it...  If not, pick a previous release that works for you and rest assured a new release will be made as soon as possible.

**Comment by obfuscurity** [comment]
We will have a `0.9.13` official release when the issues in the https://github.com/graphite-project/graphite-web/milestones/0.9.13 milestone are resolved. I want this next release as badly as everyone else, but as @steve-dave mentioned, we all have personal commitments and there are no companies sponsoring the development time of Graphite right now.

**Comment by biancalana** [comment]
There are 4 issues on  https://github.com/graphite-project/graphite-web/milestones/0.9.13, three of then already have pull request assigned (https://github.com/graphite-project/graphite-web/pull/1212, https://github.com/graphite-project/graphite-web/pull/1208 and https://github.com/graphite-project/graphite-web/pull/1157) the other one (https://github.com/graphite-project/graphite-web/issues/1081) looks solved waiting some more agreements from @SEJeff  @obfuscurity @dstufft

Anything that needs additional help ?

**Comment by hggh** [comment]
@obfuscurity I think the release should named 0.9.14, because there is already a 0.9.13 release on PyPi:
https://pypi.python.org/pypi/graphite-web/0.9.13

**Comment by obfuscurity** [comment]
That was the only way (afaict) to get an RC out there for general testing. Is there a good reason *not* to simply replace the existing version?

**Comment by hggh** [comment]
I don't know what pypi will do if you have already a 0.9.13 installed and you replace the version. Better play safe and bump the version, it's only a number :)

**Comment by dstufft** [comment]
Well, you can't replace the version. PyPI will reject a re-upload of the same filename if it's ever been uploaded to PyPI. Version numbers are free, use another one ;)

Also if you want to get the RC out, you can use an RC version number, like 0.9.13rc1.

**Comment by obfuscurity** [comment]
There was a valid reason for doing it like that, I'll just be damned if I can remember what it was. :tongue:

This should be a warning to everyone involved: get a new release out there ASAP before I get my hands on it again. :wink:

**Comment by dstufft** [comment]
It might be that pip won't install pre-releases by default, because generally people want opt-in to prereleases.

**Comment by hggh** [comment]
don't worry @obfuscurity! thanks your work on graphite. it's just a number.

**Comment by obfuscurity** [comment]
Can we get some eyes on #1209? I'd like to get this merged in.

**Comment by munhitsu** [comment]
It's 2015 - will there ever be a new graphite release?

**Comment by obfuscurity** [comment]
Working on it.

**Comment by obfuscurity** [comment]
Only a couple PRs remaining, waiting for feedback/testing: #1212 and #1295.

**Comment by tuxinaut** [comment]
@obfuscurity great :smile:

**Comment by obfuscurity** [comment]
Sorry, failed to mention also a couple of issues and PRs for Carbon too: https://github.com/graphite-project/carbon/milestones/0.9.13

**Comment by biancalana** [comment]
What about https://github.com/graphite-project/carbon/issues/366 ?
I'm facing this on everything after 0.9.12 of carbon...

**Comment by obfuscurity** [comment]
Very exciting, the graphite-web milestone is clear and the carbon milestone should be cleaned up very soon.

**Comment by torkelo** [comment]
Great news!

**Comment by JeanFred** [comment]
You’ve been busy with all these PR reviewed and merged, congrats :)

**Comment by obfuscurity** [comment]
https://github.com/graphite-project/graphite-web/pull/1349 updates the release notes.

**Comment by obfuscurity** [comment]
Our docs builds have been failing on RTD [since Aug 13](https://readthedocs.org/projects/graphite/builds/3236334/). @gwaldo is this something you can help take a look at?

**Comment by gwaldo** [comment]
Sure, I'll try to take a look today.

**Comment by gwaldo** [comment]
Looking into this for a little bit, the most plain errors are `/opt/graphite can't be created` and that django isn't being imported.

Looking at the last build to [succeed](https://readthedocs.org/projects/graphite/builds/3192585/), it looks like a very different style of information being presented.

I just signed up for a readthedocs.org account.  Does anything need to be done to add me to the RTD for Graphite so that I can see what options are available to manage it or engage with the RTD team?

**Comment by obfuscurity** [comment]
@gwaldo Added you as a project admin in RTD.

**Comment by obfuscurity** [comment]
Note that aside from the RTD docs build problem, https://github.com/graphite-project/whisper/pull/136 is our last showstopper bug before the next 0.9.x release.

**Comment by obfuscurity** [comment]
Anyone have time to look at that RTD docs build failure for 0.9.x yet? That's the only blocker at this point.

**Comment by deniszh** [comment]
It looks like I found something, but I do not know how to fix it. I'm absolute novice in RTD, so...
![NFI](https://i.kinja-img.com/gawker-media/image/upload/japbcvpavbzau9dbuaxf.jpg)

It seems that pip on RTD now honors setup.cfg settings, if it's located in same directory -  that's why when RTD tries to run
```...pip install --exists-action=w -rdocs/requirements.txt``` it tries to write to '/opt/graphite' (because we still have
```
[install]
prefix=/opt/graphite
```
in setup.cfg for 0.9.x - and fails.
@obfuscurity tried to fix that with https://github.com/graphite-project/graphite-web/commit/8539ebb1c72351a091a2825b3943acae3c1a8b87, but problem that installing docs dep's step running before running ```pip setup.py install```...
I found no ways how we can change pip behaviour, so, we need to change setup.py for 0.9.x to similar setup as we have in master. I'll create PR for that.
Please test.

**Comment by obfuscurity** [comment]
![its_fine](https://cloud.githubusercontent.com/assets/494338/10730279/613a6388-7bc5-11e5-8b46-508dfe6865dc.gif)

**Comment by torkelo** [comment]
yay!

**Comment by obfuscurity** [comment]
I don't think there's anything holding up the 0.9.14 release except for tagging and building pypi packages. I made some small improvements to the docs the other day to address what I saw was a gap for the installation process (specifically, setting up the db). We still intend to rollout a new website, but it's not ready yet, and I don't want it to hold up the new release any longer.

Aside from those things, anything else I'm overlooking?

**Comment by JeanFred** [comment]
Can’t think of anything @obfuscurity :)

**Comment by bmhatfield** [comment]
Let 'er rip!

**Comment by gwaldo** [comment]
There is a significant gut-and-rebuild of the docs that I want to do, but there's no reason to hold this up.  I'll be happy to see this release.

**Comment by obfuscurity** [comment]
I've bumped the setup.py versions, tagged the 0.9.14 release, built new source distributions, and uploaded the releases to PyPi for graphite-web, carbon, and whisper. There needs to be a formal writeup somewhere (or better yet, automated), but for the time being here are the steps I performed:

* bump the version in `setup.py`
* git tag the new version
* push both to origin
* launched a new Synthesize vagrant (this clones and checks out the 0.9.14 release)
* inside the checkout for each, `sudo python setup.py sdist`
* extract the `PKG_INFO` file from each tarball
* login to https://pypi.python.org/pypi as myself (project admin)
* for each project, submit PKG_INFO to https://pypi.python.org/pypi?%3Aaction=submit_form
* upload source distribution tarball for each release ("files")

**Comment by obfuscurity** [comment]
I would appreciate testing and feedback of the 0.9.14 pypi packages.

P.S. Note that we have recently resumed work on the new website. With this will come the shutdown of the old wikidot site, improved docs (including "quickstarts"), and other goodies. But I felt it was in the project's best interest to not hold back 0.9.14 any longer. We will aim to ship the new site ASAP, but if it happened to coincide with a new release from `master`, I think that would be cool too. :wink:

P.P.S. Thanks to everyone's hard work and persistence for helping make this release a reality.

**Comment by obfuscurity** [comment]
P.P.P.S. I've also hidden the 0.9.13-pre1 packages on PyPi.

**Comment by jraby** [comment]
Congrats!

I'll test this out next week.

**Comment by bmhatfield** [comment]
We're also looking to bring our install up to current as well, so I'll be upgrading over the next couple weeks as well.

**Comment by obfuscurity** [comment]
@brutasse Any suggestions on how to handle the db upgrade path from older fixtures to 0.9.14? For example, in Django 1.4 `syncdb` fails as such:

```
$ sudo PYTHONPATH=/opt/graphite/webapp django-admin.py syncdb --settings=graphite.settings
Creating tables ...
Creating table url_shortener_link
The following content types are stale and need to be deleted:

    auth | message

Any objects related to these content types by a foreign key will also
be deleted. Are you sure you want to delete these content types?
If you're unsure, answer 'no'.

    Type 'yes' to continue, or 'no' to cancel: yes
Installing custom SQL ...
Installing indexes ...
Problem installing fixture 'initial_data.json': Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/commands/loaddata.py", line 196, in handle
    obj.save(using=using)
  File "/usr/local/lib/python2.7/dist-packages/django/core/serializers/base.py", line 165, in save
    models.Model.save_base(self.object, using=using, raw=True)
  File "/usr/local/lib/python2.7/dist-packages/django/db/models/base.py", line 529, in save_base
    rows = manager.using(using).filter(pk=pk_val)._update(values)
  File "/usr/local/lib/python2.7/dist-packages/django/db/models/query.py", line 557, in _update
    return query.get_compiler(self.db).execute_sql(None)
  File "/usr/local/lib/python2.7/dist-packages/django/db/models/sql/compiler.py", line 986, in execute_sql
    cursor = super(SQLUpdateCompiler, self).execute_sql(result_type)
  File "/usr/local/lib/python2.7/dist-packages/django/db/models/sql/compiler.py", line 818, in execute_sql
    cursor.execute(sql, params)
  File "/usr/local/lib/python2.7/dist-packages/django/db/backends/sqlite3/base.py", line 337, in execute
    return Database.Cursor.execute(self, query, params)
IntegrityError: Could not load contenttypes.ContentType(pk=10): UNIQUE constraint failed: django_content_type.app_label, django_content_type.model
```

Upgrading to Django 1.7 offers the `migrate` command, but it doesn't fare any better:

```
$ sudo PYTHONPATH=/opt/graphite/webapp django-admin.py migrate --settings=graphite.settings
System check identified some issues:

WARNINGS:
?: (1_6.W001) Some project unittests may not execute as expected.
	HINT: Django 1.6 introduced a new default test runner. It looks like this project was generated using Django 1.5 or earlier. You should ensure your tests are all running & behaving as expected. See https://docs.djangoproject.com/en/dev/releases/1.6/#new-test-runner for more information.
Operations to perform:
  Synchronize unmigrated apps: url_shortener, account, dashboard, tagging, events
  Apply all migrations: admin, contenttypes, auth, sessions
Synchronizing apps without migrations:
  Creating tables...
    Creating table url_shortener_link
  Installing custom SQL...
  Installing indexes...
Traceback (most recent call last):
  File "/usr/local/bin/django-admin.py", line 5, in <module>
    management.execute_from_command_line()
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/__init__.py", line 385, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/__init__.py", line 377, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/base.py", line 288, in run_from_argv
    self.execute(*args, **options.__dict__)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/base.py", line 338, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/commands/migrate.py", line 128, in handle
    created_models = self.sync_apps(connection, executor.loader.unmigrated_apps)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/commands/migrate.py", line 297, in sync_apps
    call_command('loaddata', 'initial_data', verbosity=self.verbosity, database=connection.alias, skip_validation=True, app_label=app_label, hide_empty=True)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/__init__.py", line 115, in call_command
    return klass.execute(*args, **defaults)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/base.py", line 338, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/commands/loaddata.py", line 61, in handle
    self.loaddata(fixture_labels)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/commands/loaddata.py", line 91, in loaddata
    self.load_label(fixture_label)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/commands/loaddata.py", line 148, in load_label
    obj.save(using=self.using)
  File "/usr/local/lib/python2.7/dist-packages/django/core/serializers/base.py", line 173, in save
    models.Model.save_base(self.object, using=using, raw=True)
  File "/usr/local/lib/python2.7/dist-packages/django/db/models/base.py", line 618, in save_base
    updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
  File "/usr/local/lib/python2.7/dist-packages/django/db/models/base.py", line 680, in _save_table
    forced_update)
  File "/usr/local/lib/python2.7/dist-packages/django/db/models/base.py", line 724, in _do_update
    return filtered._update(values) > 0
  File "/usr/local/lib/python2.7/dist-packages/django/db/models/query.py", line 600, in _update
    return query.get_compiler(self.db).execute_sql(CURSOR)
  File "/usr/local/lib/python2.7/dist-packages/django/db/models/sql/compiler.py", line 1004, in execute_sql
    cursor = super(SQLUpdateCompiler, self).execute_sql(result_type)
  File "/usr/local/lib/python2.7/dist-packages/django/db/models/sql/compiler.py", line 786, in execute_sql
    cursor.execute(sql, params)
  File "/usr/local/lib/python2.7/dist-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python2.7/dist-packages/django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/usr/local/lib/python2.7/dist-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python2.7/dist-packages/django/db/backends/sqlite3/base.py", line 485, in execute
    return Database.Cursor.execute(self, query, params)
django.db.utils.IntegrityError: Problem installing fixture '/opt/graphite/webapp/graphite/initial_data.json': Could not load contenttypes.ContentType(pk=10): UNIQUE constraint failed: django_content_type.app_label, django_content_type.model
```

At this point even a manual "dump-and-restore" method would be acceptable.

**Comment by obfuscurity** [comment]
Note that I'm going to hide the pypi packages for 0.9.14 until a suitable upgrade path is identified.

**Comment by obfuscurity** [comment]
Disregard, it appears that the syncdb process for upgrading works fine. I'll update the docs.

**Comment by obfuscurity** [comment]
Updated the release notes for upgrading the database and including the release date (today). Re-tagged the release and generated a new source dist. Had to switch formats from tar.gz to zip for PyPI to allow the re-uploaded file (didn't realize this was a thing until now). Tested that pip install still works as expected.

**Comment by obfuscurity** [comment]
Ladies and gentlemen, Graphite 0.9.14 is officially released.

**Comment by obfuscurity** [comment]
Announced on Twitter.

<img width="474" alt="screen shot 2015-11-07 at 9 41 09 pm" src="https://cloud.githubusercontent.com/assets/494338/11018477/4b3699ee-8598-11e5-8a72-419e510d0cc5.png">

**Comment by obfuscurity** [comment]
For the sake of closing this issue out (oh dear god yes), I'm going to mark the "website" task complete even though it's clearly not. Trust me when I say we're working actively on this and I expect it to be out within the next 2-4 weeks.

**Comment by esc** [comment]
Nice!

On 8 November 2015 03:43:36 CET, Jason Dixon <notifications@github.com> wrote:
>Closed #677.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/graphite-project/graphite-web/issues/677#event-458046628

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.
